x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/MapReduceAggregator.java

```diff
@@ -0,0 +1,118 @@
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.ml.aggs.mapreduce.MapReduceValueSourceRegistry.REGISTRY_KEY;
+
+public abstract class MapReduceAggregator extends AggregatorBase {
+
+    private final List<ValuesExtractor> extractors;
+    private final AbstractMapReducer<?, ?, ?> mapReducer;
+    private final BigArrays bigArrays;
+    private final LongObjectPagedHashMap<Writeable> mapReduceContextByBucketOrdinal;
+
+    private final boolean profiling;
+
+    protected MapReduceAggregator(
+        String name,
+        AggregationContext context,
+        Aggregator parent,
+        Map<String, Object> metadata,
+        AbstractMapReducer<?, ?, ?> mapReducer,
+        List<ValuesSourceConfig> configs
+    ) throws IOException {
+        super(name, AggregatorFactories.EMPTY, context, parent, CardinalityUpperBound.NONE, metadata);
+
+        List<ValuesExtractor> extractors = configs.stream()
+            .map(c -> context.getValuesSourceRegistry().getAggregator(REGISTRY_KEY, c).build(c))
+            .collect(Collectors.toList());
+
+        this.extractors = extractors;
+        this.mapReducer = mapReducer;
+        this.bigArrays = context.bigArrays();
+        this.profiling = context.profiling();
+        this.mapReduceContextByBucketOrdinal = new LongObjectPagedHashMap<>(1, bigArrays);
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalMapReduceAggregation(name, metadata(), mapReducer, mapReducer.mapInit(bigArrays), profiling);
+    }
+
+    @Override
+    public final InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException {
+        InternalAggregation[] results = new InternalAggregation[owningBucketOrds.length];
+        for (int ordIdx = 0; ordIdx < owningBucketOrds.length; ordIdx++) {
+            results[ordIdx] = new InternalMapReduceAggregation(name, metadata(), mapReducer, getMapReduceContext(ordIdx), profiling);
+        }
+        return results;
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        return new LeafBucketCollectorBase(sub, null) {
+            @Override
+            public void collect(int doc, long owningBucketOrd) throws IOException {
+                mapReducer.map(extractors.stream().map(extractor -> {
+                    try {
+                        return extractor.collectValues(ctx, doc);
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                }), getMapReduceContext(owningBucketOrd));
```

Comment:
I am not sure where my original comment went...I guess something changed here and it became outdated :).

We should bubble up the `IOException` instead of throwing an unchecked one.